crates/bevy_ecs/src/entity/allocator.rs

```diff
@@ -0,0 +1,771 @@
+use bevy_platform_support::{
+    prelude::Vec,
+    sync::{
+        atomic::{AtomicBool, AtomicPtr, AtomicU32, AtomicU64, Ordering},
+        Arc, Weak,
+    },
+};
+use core::{mem::ManuallyDrop, num::NonZero};
+use log::warn;
+
+use crate::query::DebugCheckedUnwrap;
+
+use super::{Entity, EntitySetIterator};
+
+/// This is the item we store in the free list.
+struct Slot {
+    entity_index: AtomicU32,
+    entity_generation: AtomicU32,
+}
+
+impl Slot {
+    // TODO: could maybe make this `&mut`??
+    #[inline]
+    fn set_entity(&self, entity: Entity) {
+        self.entity_generation
+            .store(entity.generation(), Ordering::Relaxed);
+        self.entity_index.store(entity.index(), Ordering::Relaxed);
+    }
+
+    #[inline]
+    fn get_entity(&self) -> Entity {
```

Comment:
`get_entity` should be unsafe, since there's no guarantee someone sets the entity before getting it.